query-service-impl/src/main/java/org/hypertrace/core/query/service/validation/LimitValidation.java

```diff
@@ -0,0 +1,60 @@
+package org.hypertrace.core.query.service.validation;
+
+import io.grpc.Status;
+import io.reactivex.rxjava3.core.Completable;
+import javax.inject.Inject;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+import org.hypertrace.core.query.service.QueryServiceConfig;
+import org.hypertrace.core.query.service.QueryServiceConfig.LimitValidationConfig;
+import org.hypertrace.core.query.service.api.QueryRequest;
+
+@Slf4j
+class LimitValidation implements QueryValidation {
+
+  LimitValidationConfig config;
+
+  @Inject
+  LimitValidation(QueryServiceConfig queryServiceConfig) {
+    this.config = queryServiceConfig.getLimitValidationConfig();
+  }
+
+  @Override
+  public Completable validate(QueryRequest queryRequest, RequestContext requestContext) {
+
+    switch (config.getMode()) {
+      case ERROR:
+        if (isInvalidLimit(queryRequest.getLimit())) {
+          return Completable.error(
+              Status.INVALID_ARGUMENT
+                  .withDescription(generateErrorMessageForLimit(queryRequest.getLimit()))
+                  .asException());
+        }
+        return Completable.complete();
+
+      case WARN:
+        if (isInvalidLimit(queryRequest.getLimit())) {
+          log.warn(
+              generateErrorMessageForLimit(queryRequest.getLimit())
+                  + ". Allowing due to warn mode.",
+              queryRequest.getLimit(),
+              config.getMin(),
+              config.getMax());
+        }
+        return Completable.complete();
```

Comment:
Thought: was just thinking in the warning mode for invalid limit request, should we reset the boundary to configured min/max limit?

Reply:
> Thought: was just thinking in the warning mode for invalid limit request, should we reset the boundary to configured min/max limit?

I'd probably treat that as a separate mode (or even the behavior in error mode). To start, I'd like to find the places in the platform/UI firing queries with no limits or large limits to make sure no logic changes are needed there. For example, a UI page showing a list of events without limits shouldn't just be limited, it should be paged on the UI side.